colrev/ops/built_in/search_sources/ieee.py

```diff
@@ -79,28 +88,167 @@ def heuristic(cls, filename: Path, data: str) -> dict:
+        #if statement is used so that the api_key does not need to be entered every time after first insertion. But is this a proper validation of the key?
+        if len(settings ['packages']) != 24:
+            api_key = input("Please enter api key: ")
+            print (len(api_key))
+            self.review_manager.environment_manager.update_registry("packages", api_key)
+
+        settings=self.review_manager.environment_manager.load_environment_registry()
+        key = settings ['packages']
+
+        prev_record_dict_version = {}
+
+        query = colrev.ops.built_in.search_sources.xploreapi.XPLORE(key)
+        query.dataType('json')
+        query.dataFormat('object')
+        query.maximumResults(50000)
+        query.usingOpenAccess = False
+
+        parameter_methods = {}
+        parameter_methods["article_number"] = query.articleNumber
+        parameter_methods["doi"] = query.doi
+        parameter_methods["author"] = query.authorText
+        parameter_methods["isbn"] = query.isbn
+        parameter_methods["issn"] = query.issn
+        parameter_methods["publication_year"] = query.publicationYear
+        parameter_methods["queryText"] = query.queryText
+        parameter_methods["parameter"] = query.queryText
+
+        parameters = self.search_source.search_parameters
+        for key, value in parameters.items():
+            if key in parameter_methods:
+                method = parameter_methods[key]
+                method(value)
+
+        response = query.callAPI()
+        records = search_operation.review_manager.dataset.load_records_dict()
+        articles = response["articles"]
+
+        for article in articles:
+            article_id = article['article_number']
+            if article_id not in records:
+                record_dict = self.create_record_dict(article)
+                updated_record_dict = self.update_record_fields(record_dict)
+                record = colrev.record.Record(data=updated_record_dict)
+                added = ieee_feed.add_record(record=record)
+
+                if added:
+                    search_operation.review_manager.logger.info(
+                        " retrieve " + record.data["ID"]
+                    )
+                    ieee_feed.nr_added += 1
+                else:
+                    changed = self.update_existing_record(
+                    search_operation, records, record.data, prev_record_dict_version, rerun
+                    )
+                    if changed:
+                        search_operation.review_manager.logger.info(
+                                " update " + record.data["ID"]
+                        )
+                        ieee_feed.nr_changed += 1
+                        
+        ieee_feed.save_feed_file()       
+        search_operation.review_manager.dataset.save_records_dict(records=records)
+        search_operation.review_manager.dataset.add_record_changes()
+    
+    def create_record_dict(self, article):
```

Comment:
The `create_record_dict(...)` and `update_record_fields(...)` should be private methods (`__update_records_fields(...)`, ...).